### ISSUE TITLE
Learn.fit_one_cycle and memory allocation issue

### ISSUE THREAD

**Issue by miaoqiz** [jantic/DeOldify#124]
Hi,

Sorry to bother you again. 

I am experiencing a similar issue as described in this post:

https://forums.fast.ai/t/learn-fit-one-cycle-makes-me-run-out-of-memory-on-cpu-while-i-train-on-gpu/45428

The error is not unfamiliar:

---------------------------------------------

self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory

---------------------------------------------

There seems to be a fix on that post. 

---------------------------------------------
**Change**:

class TensorImageList(ImageList):
    def open(self, fn):
        return torch.load(fn, map_location='cpu').type(torch.float)

**To**:

class TensorImageList(ImageList):
    def open(self, fn):
        return Image(torch.load(fn, map_location='cpu').type(torch.float))

---------------------------------------------

When I applied it, somewhere in "get_colorize_data()" cannot get through because  the magic number not being recognized by "pickle". 

It occurred when it comes to "learn_critic.fit_one_cycle(4, 1e-4)".

I changed "num_workers" to 0 in "get_colorize_data()" and see what will happen.

Thanks

**Comment by jantic** [comment]
So question is- how much RAM do you actually have running this?  I have 128GB which may be why I've never noticed this....

**Comment by miaoqiz** [comment]
Hi, thank for the reply!

The problem may be caused by "Dataloader" in Pytorch. "fastai" probably used it for loading data. etc.

The discussion thread for this is:

https://discuss.pytorch.org/t/cpu-ram-usage-increasing-for-every-epoch/24475/15

Based on my observation, every time when you run "fit_one_cycle" or even "fit", the main memory consumption is increased until all the memory is allocated. 

Since you have a big database, so you can just call "fit_one_cycle" or "fit" once, so it is likely to be fine. For those who may utilize small database, there may be a need to run the code with multiple epochs . Each epoch adds memory. 

Thanks a lot!

**Comment by benswinney** [comment]
I hit a similar issue when running through some steps of the training. The fix for me was to set num_workers between 0 - 2 and a 96Gb swap space (on NVME). Otherwise the 32Gb of memory I had in my machine would be consumed very quickly.

**Comment by jantic** [comment]
Closing due to inactivity.

**Comment by RiverTre** [comment]
So sad, I also encountered this issue. fit_one_cycle() is all doing fine until it reach 100% point and my RAM is rapidly increasing and then out of memory caused python to crashed

**Comment by jantic** [comment]
@RiverTre I'll gladly accept pull requests that address this but at this point this repo is up primarily to share research and as I said above, there just hasn't been any activity on this for the past two years. But while we're at it, can you lay out what your system specs are- operating system and amount of RAM?

**Comment by RiverTre** [comment]
@jantic  Thank you so much for your reply. And to be brief, the problem was solved. 

Here are some details:
And about the details of the issue. My computer is Ubuntu2004 and has 32 GB RAM. I was using the notebook of ohmeow-blurr to fine tune BART-LAREGE-CNN, after some trying, I found the reason for the overwhelming  memory use is my metrics setting. When I use bert-score along with rouge to fine tune, bert-score uses a lot of memories and rouge uses less. Based on my 370000  full text of scientific articles, it really costs much memory ---- more than 20 G i think. So I tried training on 3700 and 37000 size dataset, it works fine.
